pkg/controllers/deploycontroller.go

```diff
@@ -0,0 +1,245 @@
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/klog/v2"
+	addonv1alpha1 "open-cluster-management.io/api/addon/v1alpha1"
+	addonv1alpha1lister "open-cluster-management.io/api/client/addon/listers/addon/v1alpha1"
+	clusterv1lister "open-cluster-management.io/api/client/cluster/listers/cluster/v1"
+	workv1client "open-cluster-management.io/api/client/work/clientset/versioned"
+	clusterv1 "open-cluster-management.io/api/cluster/v1"
+	workv1 "open-cluster-management.io/api/work/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+func init() {
+	utilruntime.Must(addonv1alpha1.AddToScheme(scheme))
+	utilruntime.Must(workv1.AddToScheme(scheme))
+	utilruntime.Must(clusterv1.AddToScheme(scheme))
+}
+
+const (
+	imageName              = "cluster_proxy_addon"
+	defaultImagePullPolicy = "IfNotPresent"
+	imagePullSecret        = "open-cluster-management-image-pull-credentials"
+
+	// annotationNodeSelector is key name of nodeSelector annotation synced from mch
+	annotationNodeSelector = "open-cluster-management/nodeSelector"
+)
+
+var (
+	agentImage string
+)
+
+func addFlagsForDeployController(cmd *cobra.Command) {
+	cmd.Flags().StringVar(&agentImage, "agent-image", "", "The image of agent")
+}
+
+type reconcileDeployAgentManifestwork struct {
+	client        client.Client
+	addonLister   addonv1alpha1lister.ManagedClusterAddOnLister
+	clusterLister clusterv1lister.ManagedClusterLister
+	workClient    workv1client.Interface
+}
+
+func registerDeployController(addonLister addonv1alpha1lister.ManagedClusterAddOnLister, clusterLister clusterv1lister.ManagedClusterLister, workClient workv1client.Interface, mgr manager.Manager) error {
+	c, err := controller.New("deploy-controller", mgr, controller.Options{
+		Reconciler: &reconcileDeployAgentManifestwork{
+			client:        mgr.GetClient(),
+			addonLister:   addonLister,
+			clusterLister: clusterLister,
+			workClient:    workClient,
+		},
+	})
+	if err != nil {
+		return err
+	}
+
+	if err := c.Watch(&source.Kind{Type: &addonv1alpha1.ManagedClusterAddOn{}}, &managedcluteraddonHandler{}); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (r *reconcileDeployAgentManifestwork) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
+	// get managedClusterAddon name and managedClusterNamee
+	managedClusterAddonName := request.Name
+	managedClusterName := request.Namespace
+	if len(managedClusterName) == 0 || len(managedClusterAddonName) == 0 {
+		klog.Infof("Skip reconcile: managedClusterAddonName or managedClusterName is empty")
+		return reconcile.Result{}, nil
+	}
+
+	// get managedCluster
+	managedCluster, err := r.clusterLister.Get(managedClusterName)
+	if err != nil {
```

Comment:
if not found, ignore?

Reply:
if managed cluster is deleting, ignore?

Reply:
Yes, here we skip "not found and deleting", this controller only acts as a reader.

Reply:
Fixed.